### ISSUE TITLE
Error in transform (LocalMaxPeakFinder)

### ISSUE THREAD

**Issue by ambrosejcarr** [spacetx/starfish#899]
TL;DR: to reproduce, pull my branch `ajc-osm-fish` and run `notebooks/osmFISH.ipynb`. The checked in notebook displays the expected failure, detailed below. 

I'm running into a problem with LocalMaxPeakFinder x Transform

The LocalMaxPeakFinder runs using increasing thresholds until no spots are found, at which point it compares across runs to find the right threshold. Unfortunately, when no spots are found, `ImageStack.transform` throws an exception, resulting in this component breaking. 

```python
min_distance = 6
stringency = 0
min_obj_area = 6
max_obj_area = 600

lmp = SpotFinder.LocalMaxPeakFinder(
    min_distance=min_distance,
    stringency=stringency,
    min_obj_area=min_obj_area,
    max_obj_area=max_obj_area
)
lmp_res = lmp.run(mp)
---------------------------------------------------------------------------
RemoteTraceback                           Traceback (most recent call last)
RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/usr/local/Cellar/python/3.6.5_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "/Users/ambrosecarr/projects/spacetx/starfish/starfish/imagestack/imagestack.py", line 844, in _multiprocessing_workflow
    return worker_callable(sliced, *indices_and_slice_list)
  File "/Users/ambrosecarr/projects/spacetx/starfish/starfish/util/_functools.py", line 39, in __call__
    return self.func(*called_args, **kwargs)
  File "/Users/ambrosecarr/projects/spacetx/starfish/starfish/spots/_detector/local_max_peak_finder.py", line 232, in image_to_spots
    masked_image[spot_prop.coords[:, 0], spot_prop.coords[:, 1]] = 0
IndexError: index 1 is out of bounds for axis 0 with size 1
"""

The above exception was the direct cause of the following exception:

IndexError                                Traceback (most recent call last)
<ipython-input-16-6ab3707e6d0e> in <module>
     13     max_obj_area=max_obj_area
     14 )
---> 15 lmp_res = lmp.run(mp)

~/projects/spacetx/starfish/starfish/spots/_detector/local_max_peak_finder.py in run(self, data_stack, blobs_image, reference_image_from_max_projection)
    296             reference_image_from_max_projection=reference_image_from_max_projection,
    297             measurement_function=self.measurement_function,
--> 298             radius_is_gyration=False)
    299 
    300         return intensity_table

~/projects/spacetx/starfish/starfish/spots/_detector/detect.py in detect_spots(data_stack, spot_finding_method, spot_finding_kwargs, reference_image, reference_image_from_max_projection, measurement_function, radius_is_gyration)
    234         spot_attributes_list = data_stack.transform(
    235             func=spot_finding_method,
--> 236             group_by=group_by
    237         )
    238         intensity_table = concatenate_spot_attributes_to_intensities(spot_attributes_list)

~/projects/spacetx/starfish/starfish/imagestack/imagestack.py in transform(self, func, group_by, verbose, n_processes, **kwargs)
    820             results = pool.imap(applyfunc, indices_and_slice_list)
--> 821             return list(zip(results, indices))

/usr/local/Cellar/python/3.6.5_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/pool.py in next(self, timeout)
    733         if success:
    734             return value
--> 735         raise value
    736 
    737     __next__ = next                    # XXX

IndexError: index 1 is out of bounds for axis 0 with size 1
```

In this example: 
`results = []`
`indices = [{<Indices.ROUND: 'r'>: 0, <Indices.CH: 'c'>: 0}]`

I tracked the problem down to `ImageStack.transform`, which has the following code. This function has changed a lot since I last looked at it and I'm not totally sure what's going on. I've left some questions in comments that I've added to the code, below.

```python
    def transform(
            self,
            func: Callable,
            group_by: Set[Indices]=None,
            verbose=False,
            n_processes: Optional[int]=None,
            **kwargs
    ) -> List[Any]:
        """Split the image along a set of axes, and apply a function across all the components.

        Parameters
        ----------
        func : Callable
            Function to apply. must expect a first argument which is a numpy array (see group_by)
            but may return any object type.
        group_by : Set[Indices]
            Axes to split the data along.  For instance, splitting a 2D array (axes: X, Y; size:
            3, 4) by X results in 3 arrays of size 4.  (default {Indices.ROUND, Indices.CH,
            Indices.Z})
        verbose : bool
            If True, report on the percentage completed (default = False) during processing
        n_processes : Optional[int]
            The number of processes to use for apply. If None, uses the output of os.cpu_count()
            (default = None).
        kwargs : dict
            Additional arguments to pass to func being applied

        Returns
        -------
        List[Any] :
            The results of applying func to stored image data
        """
        if group_by is None:
            group_by = {Indices.X, Indices.Y}

        indices = list(self._iter_indices(group_by))
        slice_list = [self._build_slice_list(index)[0]
                      for index in indices]

        indices_and_slice_list = zip(indices, slice_list)
        if verbose:
            indices_and_slice_list = tqdm(indices_and_slice_list)

        # why do arguments need to be stripped? what arguments are being removed here? 
        applyfunc: Callable = partial(
            self._multiprocessing_workflow,
            StripArguments(partial(func, **kwargs), positional_arguments_removed=(1, 2)),
        )

        with multiprocessing.Pool(
                n_processes,
                initializer=SharedMemory.initializer,
                initargs=((self._data._backing_mp_array,
                           self._data._data.shape,
                           self._data._data.dtype),)) as pool:
            results = pool.imap(applyfunc, indices_and_slice_list)
            # I think we need to be more careful to make sure that results is at least [[]] in the case of
            # an empty output; as it stands we can't zip these two values. 
            return list(zip(results, indices))
```

I'm not aware of all the ramifications of changing this anymore, and I guess this code isn't tested. Could one of you @ttung or @shanaxel42 take a look at this?

**Comment by ttung** [comment]
> why do arguments need to be stripped? what arguments are being removed here?

Removing that with https://github.com/spacetx/starfish/pull/906
